# Quantitative Genetics {.imageChapter}

<div class="chapter_image"><img src="../media/ch_cactus.jpg"></div>

In the vast majority of landscape genetic studies, we focus on genetic markers and how thier spatial structure is influenced by ecological, environmental, and spatial features.  This focus emphasizes neutral genetic variance, whose configuration is due to demographic and microevolutionary processes other than selection.  However, if we are interested in adaptive features in natural populations, we have a much more difficult task to accomplish, one that necessitates us to partition the neutral signal from the data so that we can identify the putatively adaptive components.  

This chapter spends a bit of time focusing on the nature of phenotypes that we can measure that may contribute to fitness. Across a landscape, the interation between gentoypes and the environment create variability in phenotypes.  These phenotypes, most commonly derived from many interacting loci, are what define the fitness landscape by the way in which they respond to differing selection pressures.

![A schematic representation of the relationship between genotypes, environmental fetures, phentoypes, and the resulting fitness landscape (adapted from Eckert & Dyer, 2012).)](../media/fitness-landscape.png)

The next chapter examines how we can look at the genetic markers alone to get inferences about where selection *may* be acting---a hypotheis generating approahc.  This one focuses on measuring and determing the extent to which traits can respond to selection.

Fisher's fundamental theorem of natural selection states that the ability of a population to respond to selection is based upon:

1. The variance of the trait.  Traits without variation cannot respond, selection has nothing to work on.  Demographic processes such as bottlenecks may create populations characterized by genetic homogeneity, which is why in conservation genetics we tend to favor practices that help to increase genetic diversity.
2. The way in which selection is acting.  If a genetic variant has a deleterious effect, the relative cost of this to the organism will be determined by the relative consequences in terms of survival and/or reproduction.  Strong effects (e.g., lethals) will obviously result in a more rapid response in the population than changes with minor effect.
3. Existing population fitness.  Populations with low average fitness may respond quicker to selection pressures than those whose average fitness is already high.

These three factors determine the extent to which selection may infleunce phenotypes in natural populations.  

## Phenotypic Variance

A phenotype is a trait that can be measured on an individual and may be morphological, biochemical, behaviorial, or any other kind of feature.  At a population level, the variance in this phenotype, $\sigma_P^2$, can be partitioned into additive components as follows:

\[
\sigma_P^2 = \sigma_G^2 + \sigma_E^2 + \sigma_{GxE}^2 
\]

The  $\sigma_G^2$ term is the proportion of the phenotypic variance that can be attributed to genetic variation.  Phenotypes are also influenced by environmental variation, $\sigma_E^2$.  Finally, there may be an interaction between the genetic and environemtnal variance, $\sigma_{GxE}^2$, that also contributes to phenotypic variance.

The genetic variance, $\sigma_G^2$, can be further partitioned into additive components salient to understanding the evolution of traits.  

\[
\sigma_G^2 =\sigma_A^2 + \sigma_D^2 + \sigma_I^2 
\]

These components are infleunced by the way in which alleles, loci, and other genetic components contribute to the production of a particular trait.

- Additive genetic variance, $\sigma_G^2$:  Additive genetic variance is created by the ineraction among alleles at a particluar locus as well as the contributions of loci in the development of a trait.  In the simplest model, consider a locus that encodes for a protein that modifies a substrate component in the biochemical pathway for a phenotype (say pigmentation development that leads to flower color).  Variant alleles may be more or less efficient in their capacity to modify substrate causing differences among gentypes.  Loci along this pigmentation pathway may individually have many variant types, all of which contribute to the resulting phenotype.  This is additive variance.  
- Dominance variance, $\sigma_D^2$:  Dominance variance can be similarly defined based upon the interaction among alleles at a locus.  In the classic Mendelian model, a dominant allele can mask the effects of other alleles. This deviates from the pure additive model in that all you need at a diploid locus is a single copy of the dominant allele to express the dominant pheotype.
- Interaction variance, $\sigma_I^2$:  Genetic interaction variance captures the way in which alternative loci influence each other.  The classic case here is that of coat color in the Labrador dog breed.  One locus determines the pigmentation color, with alleles that encode for black pigmentation color dominant to brown.  An entirely different locus determines the ability to incorporate the pigmentation into the hair shaft itself with a dominant allele that allows the pigentation to be incorporated (resulting in either a black or brown dog) and a recessive allele the prevents incorporation (producing the yellow labrador).  At each locus, the interaction of the alleles contribute to the dominance variance but the way in which the second locus infleunces the phenotype produced by the first locus is interaction variance.

Estimating the features that contribute to a phenotype is not trivial and  requires specific sampling approahces which often include analyses across generations and may require precise information on the relatedness of individuals or controlled matings and methods to standardize environmental variance.  

## Data Examples

Here we will use some example data that was collected to understand the quantitative consequences of urban cultivar gene escape on native populations in the tree *Cornus florida*.  This species is widley distributed in natural populations throughout forests of easter North America.  It has also been artifically selected to express traits that make it suitable for use in the cultivar trade by nurseries.  Cultivar types can easily thrive in open sunlight (as is often found in residental yards, etc), whereas the native variants are typically typically found in more shaded areas of the forest understory.

An experiment was set up to collect arrays of seedling siblings from maternal individuals in both the urban environment and proximate native populations.  These seeds were germinated and grown in a greenhouse environment. Seedlings were then randomly assigned to either a full light (to mimic the urban environment) or a shaded (to represent the native understory) environment.  Several quantiative measurements were collected throughout the growning season.  A subset of these data are available in the <tt>seedlings.rda</tt> data file.


![Dogwood seedlings growing in the greenhouse.  Seeds were collected from parental trees from urban (left) and natural (right) populations and grown in either full light (to mimic urban environments) and shaded (to mimic native popualtions)](./media/seedlings.png)

```{r echo=FALSE}
load("media/seedlings.rda")
t <- table( seedlings$Source, seedlings$Mom)
df <- as.data.frame( t(t) )
df <- df[ !df$Freq == 0,]
```

The data presented represent the seedlings that had survived to the end of the growing season, partitioned by seed source.  A summary of the data is listed below.

Source | Treatment | Number of Trees | Offspring
-------|:----------|:---------------:|:----------:
fan    | Cultivar  | 24              | 131
rice   | Native    | 9               | 40

Quantitative measurements were made on each seedling throughout the growing season to determine if cultivar types were producing different phenotypes than native stock.  Some of the measured traits include:

- Cold - The length of time (in days) that the seedling remained dormant after cold stratification.
- Branches - The number of branching stems on each seedling.
- Leaves - The total number of leaves on the seedling.  
- Height - Total seedling height.
- Leaf Area - The size of the largest leaf on the seedling.

```{r}
summary(seedlings)
```

The relationship among these traits are presented below.

```{r message=FALSE, warning=FALSE}
library(GGally)
library(ggplot2)
ggpairs( seedlings, columns=4:8)
```

Like most data, these are imperfect.  In particular, there are some families whose seedling mortality resulted in few seeds.

```{r}
t <- table( seedlings$Mom )
t
```

As a result, before we go forward, lets limit the data to those mothers whose families have  at least 5 surviving offspring.

```{r}
keepers <- names(t)[ t >= 4 ] 
seedlings <- seedlings[ seedlings$Mom %in% keepers, ]
seedlings <- droplevels( seedlings )
```



### Measured Traits

Visual summaries of the measured traits are shown below.  

```{r fig.cap="Length of time (in days) of seedling dormancy following cold stratification for seedlings collected from maternal trees in urban (fan) and native (rice) populations."}
p <- ggplot(seedlings) + theme_bw()
p + geom_density( aes(fill=Source,x=Cold), alpha=0.75)
```

In the remaining plot, I partion the display of phenotypic variability further by separating out seedlings that were grown in either the SUN (full light) or SHADE (reduced light) treatments.

```{r fig.cap="The height of the seedling measured from the soil to the top of the apical meristem."}
p <- p + facet_grid(Treatment~.)
p + geom_density( aes(fill=Source,x=Height), alpha=0.75) 
```

```{r fig.cap="The number of branches from the main seedling stem."}
p + geom_density( aes(fill=Source,x=Branches), alpha=0.75) 
```

```{r fig.cap="The photosynthetic area measured on the largest leaf from each seedling."}
p + geom_density( aes(fill=Source,x=LeafArea), alpha=0.75) 
```

```{r fig.cap="The total number of leaves on the seedling."}
geom_density( aes(fill=Source,x=Leaves), alpha=0.75)
```

Overall, there are noticable differences in phenotypes due to both `Source` and `Treatment` as well as some pronounced bi-modality in the data.


## Heritability

For a phenotype to be influenced by natural selection, it must be heritable.  Heritability is defined as the extent to which quantitative variation in inherited across generations.  A distinction should be made between inheritance and heritability.  Inheritance is the ability to pass along a phenotype to the next generation--any phenotype.  Heritability is the passing along of phenotypic variance.  As such, all traits that have heritability are also inherited but not all inherited traits have heritability!

Heritability can be estimated at two different levels of specificity.  Broad sense heritability ($H^2$) is the proportion of the phenotypic variance that can be attribted to genetic variance.

\[
H^2 = \frac{\sigma_G^2}{\sigma_P^2}
\]

This estimate includes the additive ($\sigma_A^2$), dominance ($\sigma_D^2$), and interaction ($\sigma_I^2$) variance and is the easiest parameter to estimate.  A common approach to estimating $H^2$ is to report it as the proportion of variance explained ($R^2$) from an ANOVA based upon family-level data.  Often we treat the family unit as a random effect in an analysis of variance.  In `R` we fit random effects models using the `lmer` function.

```{r warning=FALSE, message=FALSE}
library(lme4)
h2.cold <- lmer(Cold ~ (1|Mom), data=seedlings)
summary(h2.cold)
```

If you are not familiar with fitting a random effect to a model in R, this may look a bit odd.  The formula `Cold ~ (1|Mom)` is examining differences among the length of dormancy as explained by seedlings grouped by mother.  The maternal stratum is treated as a random effect because we are assuming that the mothers we used are a random selection from a much broader population of mothers.  The differences between a random and a fixed effects model is in how we estimate the variance components.  In a fixed effect model, the expectation for the treatment variance is based solely upon the treatment effect $E[\sigma_A^2] = MSA$.  However, in a random effects model, we assume that since we are sampling on a random subsample of potential mothers, the treatment effect is a bit upwardly biased and we need to correct that variance.  

Here is an example of how those differ.  Treating mother as a fixed effect, we can see the estimates of the varince for both Mom and Residuals as the values int he third column.

```{r}
anova( aov(Cold~Mom,data=seedlings))
```

However, those same variance estiamtes, assuming that Mom is a random effect, are different (moreso for the Mom effect than that for the Residual variance).

```{r}
print(VarCorr(h2.cold), comp="Variance")
```

Then from the formula above, we can estimate the fraction of the total variance that is attributed to the family effect as:

```{r}
sigmas <- as.data.frame( VarCorr( h2.cold) )$vcov
H2 <- sigmas[1] / sum( sigmas )
H2
```

This result shows suggests that roughly `r round(H2*100,digits=2)`% of the variation can be attributed to genetic effects.

Narrow-sense heritabilty is a bit more precise in that it uses not the total genetic variance but only that associated with the additive component of variation.  

\[
h^2 = \frac{\sigma_A^2}{\sigma_P^2}
\]

However, this precision is gained at the expense of a bit more complexity.  There are a few different basic methods used for estimating $h^2$ from sibling data.

1. Twin studies.  If you have data from monozygotic (MZ) and dizygotic (DZ) twins, the narrow sense heritability can be estimated using the differenes in phenotypic correlations, $h^2 = r_{MZ} - r_{DZ}$ (the so-called Falconer's Formula).  Jaccard (1983) had suggested a modification of this to use $h^2 = (r_{MZ} - r_{DZ})/(1-r_{DZ})$ instead though it is not that common of a standardization.  
2. Another method that can be used relies upon estimates of consanguanity (e.g., genetic relatedness) of the siblings.  In most organisms, this would be a value that is relatively easy to calculate---they are half sibs because we collected them from the same maternal individual!  However, in many plants it is not uncommon to have mixes of full and half siblings due to selfing as well as have correlated paternity (e.g., two seeds are actually full siblings since they share the same fater as well as mother).  Without a little more genetic information on these individuals we cannot determine either of these parameters.
3. The final method uses both parent and offspring phenotypes.  Here, narrow sense heritabilty is the slope of a regression line fitting offspring phenotype to the average of the parental phenotype.  If heritabilty is high, then the slope of this line will approach 1.0, whereas if offspring phenotypes have little relation to those measured on the parents (e.g., $\sigma_E^2$ and perhaps $\sigma_{GxE}^2$ have more influence) then the slope will tend towards zero.  There is a caveat here in that when a trait has a low heritability (e.g., the slope is close to zero), it is often the case that the variation explained by that model (the $R^2$) is also low.  Conversely, when heritability is high, it is often the case that $R^2$ is also increased.  This means that the power of this kind of anlaysis improves with heritability in that we get a better estimate when it is high.

You can easily find examples of each of these methods in the literature.  However, in landscape studies, we often lack the kind of information necessary to collect these data, which is why understanding quantitative genetics in natural populations is a challenging endeavour.  Unfortunately, the data we are working with do not have the additional relatedness information or parental seedling characteristics included so we cannot estimate narrow sense heritibility.  

## Questions

1. For many quatitative traits, environmental variance ($\sigma_E^2$) may make large contributions to observed phenotypic variance.  From the traits above, which show the largest environmental effects (e.g., `SHADE` vs `SUN`)?  

```{r echo=FALSE, eval=FALSE}
The branches trait seems to be the most influenced by SHADE vs. SUN.  There is very little variance in the SHADE treatment but there is a broad range of values for the SUN one.
```


2. What is the broad-sense heritability for the "Height" and the "Leaf Area" traits in the `SHADE` treatment?  (Make sure to drop any family that has too few seedlings) Are they similar or different?  Explain the results.

```{r eval=FALSE, echo=FALSE}

data <- seedlings[ seedlings$Treatment=="SHADE",]
t <- table( data$Mom) 
keepers <- names(t)[ t >= 4 ] 
data <- data[ data$Mom %in% keepers, ]
data <- droplevels( data )

t <- lmer(Height ~ (1|Mom), data=data)
sigmas <- as.data.frame( VarCorr( t) )$vcov
H2.height <- sigmas[1] / sum( sigmas )
cat("H2 for height",s,"\n")

t <- lmer(LeafArea ~ (1|Mom), data=data)
sigmas <- as.data.frame( VarCorr( t) )$vcov
s <- sigmas[1] / sum( sigmas )
cat( "H2 for leaf area:", s, "\n")
```


3. Estimate broad-sense heritability for the `SUN` families for "Height" and "Leaf Area".  How do these differ from the estimates of heritability in the `SHADE` seedlings?

```{r eval=FALSE, echo=FALSE}
data <- seedlings[ seedlings$Treatment=="SUN",]
t <- table( data$Mom) 
keepers <- names(t)[ t >= 4 ] 
data <- data[ data$Mom %in% keepers, ]
data <- droplevels( data )

t <- lmer(Height ~ (1|Mom), data=data)
sigmas <- as.data.frame( VarCorr( t) )$vcov
H2.height <- sigmas[1] / sum( sigmas )
cat("H2 for height",s,"\n")

t <- lmer(LeafArea ~ (1|Mom), data=data)
sigmas <- as.data.frame( VarCorr( t) )$vcov
s <- sigmas[1] / sum( sigmas )
cat( "H2 for leaf area:", s, "\n")
```


4. Assuming that the dominance and interaction variance ($\sigma_D^2$ and $\sigma_I^2$) of the height and leaf area traits were constant (e.g., $H^2$ would be proportional to $h^2$), rank the traits in terms of their ability to respond to selection.  Which are able to respond faster?  

```{r eval=FALSE, echo=FALSE}
Traits with the largest heritability will be the ones that respond the quickest to any type of selection, assuming that the strength of selection (e.g., the mean value of the parents relative to that of the population they are drawn from) is the same.
```

5. What are some of the challenges associated with this partiular data set?

```{r eval=FALSE, echo=FALSE}
- low survivalship
```

